src/MCPClient/lib/clientScripts/pid_declaration.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python2
+# -*- coding: utf8
+
+"""pid_declaration.py
+
+Given an identifiers.json file, supplying third-party persistent identifiers
+(PIDS), such as handle (hdl) or doi, identifiers; associate those identifiers
+with the objects in the transfer, so be translated to PREMIS objects in the
+AIP METS.
+"""
+from __future__ import unicode_literals
+from functools import wraps
+import json
+import os
+import sys
+
+import django
+
+django.setup()
+
+from main.models import Directory, File, SIP
+
+from pid_component_helpers import (
+    _add_custom_pid_to_mdl_identifiers,
+    AssignPIDsException,
+)
+from sanitize_names import sanitizeName
+
+
+class DeclarePIDsException(Exception):
+    """Exception to raise if there's a more fundamental issue with the running
+    of this script and we want to exit and stop the workflow from continuing.
+    """
+
+
+class DeclarePIDsWarning(Exception):
```

Comment:
I'm not totally clear on the behaviour desired — it seems like we want to short circuit the script but still exit with success? If that's the case it seems like we don't need the warning and context manager, we can just add an except for the error in `call`.

I find the naming on this a bit confusing, if something is named `Warning`, I'd expect something inheriting from [the builtin Warning](https://docs.python.org/2/library/warnings.html).

Reply:
I wonder if I am missing something, but if I add the `except` we can't prevent the code from exiting when the exception is first raised. I haven't managed to return an exception back to `call` yet. 

The model we've adopted seems to have evolved over time, with some of our scripts trying to do a lot of scripts doing strange wrangling with states and asking permission. 

This is my first time trying this, adopting Joel D's approach in the similar bind_pids work: https://github.com/artefactual/archivematica/blob/84661775836cf8037ad3b48feb8e02bb80335f0f/src/MCPClient/lib/clientScripts/bind_pids.py#L75-L85 

But yeah, this just allows us to short-circuit as you say, from any point in the code and return back to the MCP server with a non-positive status if we're just trying to exit when something's non-critical. 

If you can help me get it to work using an `except` in call that would be great.